### DISCUSSION TITLE
Trying to build custom extension for USB4/Thunderbolt

### DISCUSSION THREAD

**Discussion by ajgon** [siderolabs/talos#7508] [Q&A] [answered]
Hey, I'm trying to build my custom extensions for `thunderbolt` and `thunderbolt_net` modules (CONFIG_USB4 in newer kernels). Here are my pkgs and extensions

- https://github.com/siderolabs/pkgs/compare/main...ajgon:talos-pkgs:main
- https://github.com/siderolabs/extensions/compare/main...ajgon:talos-extensions:main

I'm building them this way:

```bash
# pkgs
make thunderbolt-pkg PLATFORM=amd64 USERNAME=ajgon PUSH=true
# extensions
make thunderbolt PLATFORM=amd64 USERNAME=ajgon PUSH=true
```

and eventually ending up with:

- https://github.com/users/ajgon/packages/container/package/thunderbolt-pkg
- https://github.com/users/ajgon/packages/container/package/thunderbolt

Then I'm installing extension to mine machine (only relevant part of the config):

```yaml
machine:
  kernel:
    modules:
      - name: thunderbolt_net

  install:
    image: ghcr.io/siderolabs/installer:v1.5.0-alpha.2-4-g5e13cafe5
    extensions:
      - image: ghcr.io/ajgon/thunderbolt:v1.5.0-alpha.2-dirty
```

So far so good, the modules are extracted and installed properly. Hower when I'm loading them, I'm getting

```dmesg
[    2.548858] [talos] [initramfs] enabling system extension thunderbolt v1.5.0-alpha.2-dirty
...
[    7.765646] Loading of module with unavailable key is rejected
[    7.769070] Loading of module with unavailable key is rejected
[    7.769833] [talos] controller failed {"component": "controller-runtime", "controller": "runtime.KernelModuleSpecController", "error": "error loading mod
ule \"thunderbolt_net\": load thunderbolt_net failed: key was rejected by service"}
```

and I'm bit lost. I based my module on [usb-modem-drivers pkg](https://github.com/siderolabs/pkgs/tree/main/usb-modem-drivers) and (usb-modem-drivers extension](https://github.com/siderolabs/extensions/tree/main/drivers/usb-modem) - and usb-modem is working for me. Since thunderbolt is basically a copy-paste of that approach, I assume I'm doing something wrong on build/using Makefile phase. 

`modinfo` shows that modules is signed (if I understand it correctly):

```
filename:       /host/lib/modules/6.1.39-talos/kernel/drivers/net/thunderbolt-net.ko
license:        GPL v2
description:    Thunderbolt/USB4 network driver
author:         Mika Westerberg <mika.westerberg@linux.intel.com>
author:         Michael Jamet <michael.jamet@intel.com>
author:         Amir Levy <amir.jer.levy@intel.com>
srcversion:     BCE402647E954884C7E42F3
alias:          tbsvc:knetworkp00000001v*r*
depends:        thunderbolt
retpoline:      Y
intree:         Y
name:           thunderbolt_net
vermagic:       6.1.39-talos SMP preempt modversions
sig_id:         PKCS#7
signer:         Build time throw-away kernel key
sig_key:        2D:AA:9F:B4:42:94:7A:D4:5D:B8:77:14:94:D3:80:E5:73:A3:29:54
sig_hashalgo:   sha512
signature:      9E:21:3A:2B:0A:36:85:5D:04:20:92:1E:E3:1B:E1:DC:0C:F0:9F:57:
                EB:77:1F:1B:23:EB:71:E8:A6:40:10:A9:A8:D2:44:98:6E:76:43:D4:
                7B:39:0D:B3:E8:A3:A1:8D:D1:28:62:38:B7:99:95:50:97:1F:54:81:
                2A:05:0D:42:F2:7D:5C:E2:E5:2D:C9:6D:03:E3:8A:08:9E:20:0E:B2:
                2E:D0:61:D0:3F:67:8C:7E:91:62:D8:47:F0:64:9B:CD:9C:25:F6:A2:
                CC:40:89:03:21:B5:04:31:BE:1F:22:55:CE:A4:9B:87:13:60:AA:23:
                6F:C4:CC:D9:86:16:63:78:0B:AE:90:CB:76:62:EF:4D:75:43:3D:21:
                AA:76:BF:2E:D7:AB:F0:89:42:BA:DD:F3:18:C2:A9:CD:82:B9:97:95:
                83:F9:8C:63:FC:53:31:67:23:E3:91:B9:BC:37:71:81:DF:21:07:18:
                D4:68:A9:36:34:C3:64:2A:C8:97:36:88:B3:E0:E8:DF:13:DA:16:F2:
                AC:C3:4A:79:E7:BC:FC:A1:40:A9:75:A8:6D:05:6F:7E:1A:18:1D:3A:
                F2:A1:8D:D4:43:30:3A:96:94:44:B9:15:CD:7A:72:CB:37:2F:2C:BF:
                DD:9A:FF:37:22:F0:28:B0:6C:BE:A6:ED:A0:7F:A7:7E:80:F6:51:17:
                75:9D:B8:BC:98:C4:4F:34:65:24:3C:F4:3D:F3:A1:B6:AA:9A:A3:99:
                B6:F2:77:5F:91:95:65:3D:02:06:FD:9B:2E:53:1B:F5:A1:75:C2:23:
                04:6D:32:BB:A1:C4:94:92:FC:56:B6:65:45:E7:A5:D1:E1:6D:93:F6:
                1A:A5:6A:F6:2E:62:E1:CD:E4:3C:98:29:C6:CE:A5:F6:A9:AF:BA:F7:
                62:E8:D7:53:FA:F0:A0:70:AB:FA:95:3F:9D:05:0C:D4:C3:D4:00:76:
                3A:26:6E:0C:96:1D:E2:6A:3E:B6:D9:9B:2E:76:B4:B8:96:99:15:55:
                AE:40:27:3E:8A:F2:74:18:3E:22:55:81:B6:CB:4C:93:D0:A9:F0:79:
                93:9B:CE:D9:0F:2E:42:30:AC:AF:A9:2C:AB:25:C2:88:7E:53:A8:29:
                11:03:51:B8:64:15:D6:61:E7:DA:10:EA:DF:21:9D:35:AA:23:EA:1D:
                1F:C9:76:F6:A2:34:4E:C9:AC:26:17:73:5D:16:A0:84:93:83:8D:6B:
                F8:F6:53:76:0E:C9:0E:E7:E4:97:70:B0:C7:7E:40:9E:EF:3A:AD:64:
                F3:8C:07:C3:A3:41:80:C5:09:2D:F7:7D:3F:4F:F5:EE:6A:0E:D6:A3:
                E6:9A:84:8A:08:91:D8:3F:A4:84:D0:83
parm:           e2e:USB4NET full end-to-end flow control (default: true) (bool)
```

Also this machine has secureboot disabled.

At this point, I'm out of ideas, and any hint will be greatly appreciated.

**Comment by frezbo** [comment]
This is probably due to buildkit being aggressive with dropping caches, would recommend to create a builder with a custom config like here: https://github.com/siderolabs/build-container/blob/main/hack/scripts/setup-ci#L12

https://github.com/siderolabs/build-container/blob/main/hack/scripts/buildkit.toml

**Reply by ajgon** [reply]
Unfortunatelly it didn't help. I used the provided `buildkit.toml` and created builder this way:

```bash
docker buildx create --name mybuild --buildkitd-flags "--allow-insecure-entitlement security.insecure" --driver-opt image="docker.io/moby/buildkit:v0.11.6" --platform linux/amd64 --driver docker-container --config ./buildkit.toml --use
```

And double checked, that new image was built using this builder. I also dropped all caches before-head to start clean. Still no luck :(

**Comment by ajgon** [comment, marked as answer]
Found the issue, let me answer for anyone stumbling into this:

The problem was, that I was using default talos installer image with kernel built by talos team, and with random signing key used by it. So when I introduced my image with thunderbolt module, naturally I didn't have this key and another one was created - thus creating divergence.

After I built my own kernel and my own installer using [this guide](https://www.talos.dev/v1.4/advanced/customizing-the-kernel/) everything clicked.

P.S. I think the kernel building guide has a mistake, the following line didn't work for me:

```dockerfile
COPY --from=<custom kernel image> /lib/modules /lib/modules
```

had to replace it to 

```dockerfile
COPY --from=<custom kernel image> /lib/modules /kernel/lib/modules
```

to make everything work.

**Reply by frezbo** [reply]
> P.S. I think the kernel building guide has a mistake, the following line didn't work for me:

> COPY --from=<custom kernel image> /lib/modules /lib/modules

> had to replace it to

> COPY --from=<custom kernel image> /lib/modules /kernel/lib/modules


This is a small interim change for 1.5 (would have a better solution before release)

The easy way to test is to edit the dockerfile of Talos repo, point to the custom kernel image and run `make installer PUSH=true`
